<h2><font color="#004D7F" size=6>Módulo 2. Análisis de datos</font></h2>



<h1><font color="#004D7F" size=5>2. Estadística descriptiva</font></h1>

<br><br>
<div style="text-align: right">
<font color="#004D7F" size=3>Manuel Castillo-Cara</font><br>
<font color="#004D7F" size=3>Machine Learning con Python</font><br>

---

<h2><font color="#004D7F" size=5>Índice</font></h2>
<a id="indice"></a>

* [1. Introducción](#section1)
    * [1.1. Cargar el conjunto de datos](#section11)
* [2. Funciones de estadística descriptiva](#section2)
    * [2.1. Revisar los datos: _head()_](#section21)
    * [2.2. Dimensiones de los datos: _shape_](#section22)
    * [2.3. Tipo de datos: _dtypes_](#section23)
    * [2.4. Resumen: _describe()_](#section24)
    * [2.5. Distribución entre clases: _groupby('class').size()_](#section25)
    * [2.6. Correlaciones: _corr()_](#section26)
    * [2.7. Asimetría: _skew()_](#section27)

In [ ]:
# Permite ajustar la anchura de la parte útil de la libreta (reduce los márgenes)
from IPython.core.display import display, HTML
display(HTML("<style>.container{ width:98% }</style>"))

---

<a id="section1"></a>
# <font color="#004D7F"> 1. Introducción</font>

Unos de los errores principales que se cometen cuando se comienza a trabajar con machine learning es tomar decisiones directamente a través de los algoritmos sin un análisis previo del conjunto de datos a trabajar. Es importante que entendamos que, un análisis de datos y un buen preprocesamiento de los mismos antes de realizar un tratamiento de modelado, nos llevará no solamente a obtener mejores y más confiables resultados, sino que entenderemos a la perfección nuestro conjunto de datos dando una gran experiencia en el tiempo en la fase de análisis y tratamiento de datos.

<a id="section11"></a>
## <font color="#004D7F">1.1. Cargar el conjunto de datos</font>

Para esta práctica vamos a cargar el conjunto de datos de Pima Indian Diabetes para hacer observaciones con las funciones de estadística descriptiva.

In [1]:
import pandas as pd 
filename='./data/pima-indians-diabetes.csv'
names=['preg','plas','pres','skin','test','mass','pedi','age','class']
data=pd.read_csv(filename,names=names)


<div style="text-align: right"> <font size=5>
    <a href="#indice"><i class="fa fa-arrow-circle-up" aria-hidden="true" style="color:#004D7F"></i></a>
</font></div>

---

<a id="section2"></a>
# <font color="#004D7F"> 2. Funciones de estadística descriptiva</font>

La inspección de nuestro conjunto de datos es fundamental para poder entender mejor que técnica utilizar; además, nos ayudará a desarrollar nuestra intuición y hacernos preguntas sobre ellos. Las múltiples perspectivas de sus datos lo desafiarán a pensar en los datos de manera diferente.

<a id="section21"></a>
## <font color="#004D7F">2.1. Revisar los datos: _head()_ </font>

Puede revisar las primeras 20 filas de sus datos utilizando la función `head()` en el DataFrame de Pandas. Puede ver que la primera columna enumera el número de fila, lo cual es útil para hacer referencia a una observación específica.

In [7]:
data.head(15)
# data.describe()

preg        plas        pres        skin        test        mass  \
count  768.000000  768.000000  768.000000  768.000000  768.000000  768.000000   
mean     3.845052  120.894531   69.105469   20.536458   79.799479   31.992578   
std      3.369578   31.972618   19.355807   15.952218  115.244002    7.884160   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      1.000000   99.000000   62.000000    0.000000    0.000000   27.300000   
50%      3.000000  117.000000   72.000000   23.000000   30.500000   32.000000   
75%      6.000000  140.250000   80.000000   32.000000  127.250000   36.600000   
max     17.000000  199.000000  122.000000   99.000000  846.000000   67.100000   

              pedi         age       class  
count   768.000000  768.000000  768.000000  
mean    428.235091   33.240885    0.348958  
std     340.485655   11.760232    0.476951  
min       0.100000   21.000000    0.000000  
25%     205.000000   24.000000    0.000000  
50%     337.000000   29.000000    0.000000  
75%     591.500000   41.000000    1.000000  
max    2329.000000   81.000000    1.000000

<a id="section22"></a>
## <font color="#004D7F">2.2. Dimensiones de los datos: _shape_ </font>

Puede revisar la forma y el tamaño de su conjunto de datos imprimiendo la propiedad `shape` en el DataFrame de Pandas. Los resultados se enumeran en filas y luego en columnas. Puede ver que el conjunto de datos tiene 768 filas y 9 columnas

In [8]:
# Dimensions of your data
data.shape

(768, 9)

<a id="section23"></a>
## <font color="#004D7F">2.3. Tipo de datos: _dtypes_ </font>

Puede enumerar los tipos de datos utilizados por el DataFrame para caracterizar cada atributo utilizando la propiedad `dtypes`. Puede ver que la mayoría de los atributos son enteros y que `mass` y `pedi` son tipos de coma flotante.

In [11]:
# Data Types for Each Attribute
data.dtypes

preg       int64
plas       int64
pres       int64
skin       int64
test       int64
mass     float64
pedi     float64
age        int64
class      int64
dtype: object

<a id="section24"></a>
## <font color="#004D7F">2.4. Resumen: _describe()_ </font>

Puedes ver que obtienes muchos datos. Notará algunas llamadas como `pandas.set_option()` utilizada para cambiar la precisión de los números y el ancho preferido de la salida. Esto es para que sea más legible para este ejemplo. Al describir sus datos de esta manera, vale la pena tomarse un tiempo y revisar las observaciones de los resultados.

In [13]:
# Statistical Summary

pd.set_option('display.width', 100) # En la nueva versión es con "display"
pd.set_option('display.precision', 3) # En la nueva versión es con "display"
data.describe()

preg     plas     pres     skin     test     mass      pedi      age    class
count  768.000  768.000  768.000  768.000  768.000  768.000   768.000  768.000  768.000
mean     3.845  120.895   69.105   20.536   79.799   31.993   428.235   33.241    0.349
std      3.370   31.973   19.356   15.952  115.244    7.884   340.486   11.760    0.477
min      0.000    0.000    0.000    0.000    0.000    0.000     0.100   21.000    0.000
25%      1.000   99.000   62.000    0.000    0.000   27.300   205.000   24.000    0.000
50%      3.000  117.000   72.000   23.000   30.500   32.000   337.000   29.000    0.000
75%      6.000  140.250   80.000   32.000  127.250   36.600   591.500   41.000    1.000
max     17.000  199.000  122.000   99.000  846.000   67.100  2329.000   81.000    1.000

<a id="section25"></a>
## <font color="#004D7F">2.5. Distribución entre clases: _groupby('class').size()_ </font>

Puede ver que hay casi el doble de observaciones con la clase 0 (sin aparición de diabetes) que con la clase 1 (aparición de diabetes). En este caso podemos observar que las clases se encuentran desbalanceadas por lo que los tenemos que analizar muy los resultados de los algoritmos.

In [20]:
data['class'].value_counts()
data.groupby('class').size()
data.groupby('class').mean()


preg     plas    pres    skin     test    mass     pedi     age
class                                                                  
0      3.298  109.980  68.184  19.664   68.792  30.304  384.240  31.190
1      4.866  141.257  70.825  22.164  100.336  35.143  510.316  37.067

<a id="section26"></a>
## <font color="#004D7F">2.6. Correlaciones: _corr()_ </font>

Puede usar la función `corr()` para calcular una matriz de correlación. La matriz enumera todos los atributos en la parte superior y lateral, para dar correlación entre todos los pares de atributos (dos veces, porque la matriz es simétrica). Puede ver que la línea diagonal a través de la matriz desde las esquinas superior izquierda a inferior derecha de la matriz muestra una correlación perfecta de cada atributo consigo mismo.

In [21]:
# Pairwise Pearson correlations
pd.set_option('display.width', 100) # con display
correlation = data.corr(method='pearson')
print(correlation)

        preg   plas   pres   skin   test   mass   pedi    age  class
preg   1.000  0.129  0.141 -0.082 -0.074  0.018 -0.026  0.544  0.222
plas   0.129  1.000  0.153  0.057  0.331  0.221  0.133  0.264  0.467
pres   0.141  0.153  1.000  0.207  0.089  0.282  0.051  0.240  0.065
skin  -0.082  0.057  0.207  1.000  0.437  0.393  0.154 -0.114  0.075
test  -0.074  0.331  0.089  0.437  1.000  0.198  0.185 -0.042  0.131
mass   0.018  0.221  0.282  0.393  0.198  1.000  0.104  0.036  0.293
pedi  -0.026  0.133  0.051  0.154  0.185  0.104  1.000  0.018  0.177
age    0.544  0.264  0.240 -0.114 -0.042  0.036  0.018  1.000  0.238
class  0.222  0.467  0.065  0.075  0.131  0.293  0.177  0.238  1.000


<a id="section27"></a>
## <font color="#004D7F">2.7. Asimetría: _skew()_ </font>

Puede calcular el sesgo de cada atributo utilizando la función `skew()`. El resultado de inclinación muestra una inclinación positiva (derecha) o negativa (izquierda). Los valores más cercanos a cero muestran menos sesgo.

In [22]:
# Skew for each attribute
data.skew()

preg     0.902
plas     0.174
pres    -1.844
skin     0.109
test     2.272
mass    -0.429
pedi     1.562
age      1.130
class    0.635
dtype: float64

<div style="text-align: right"> <font size=5>
    <a href="#indice"><i class="fa fa-arrow-circle-up" aria-hidden="true" style="color:#004D7F"></i></a>
</font></div>

---

<div style="text-align: right"> <font size=6><i class="fa fa-coffee" aria-hidden="true" style="color:#004D7F"></i> </font></div>